### PR TITLE
Fix calculating wide path status

### LINKED ISSUE #22824: Some path configurations impassable for guests due to wide path flags
### Operating System

Windows 10

### OpenRCT2 build

v0.4.14 (but has been in game for a loooooong time)

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

This is a development issue

### Describe the issue

The game sets a wide path flag for path tiles when a path is wider than 1 tile, to allow easier pathfinding for guests. In vanilla RCT2, this was done in a grid pattern. In OpenRCT2, more tiles get a wide path flag which decreases the amount of junctions. However, it does seem to lead to some situations where the guests will not be able to pathfind past a junction at all. I have encountered several situations like these when playing scenarios which leads to lost guests, which in turn kills your park rating. Here are some examples (white is impassable tiles):

pathfinding1
pathfinding2
pathfinding3
pathfinding4

Next one is not impassable, but it's a situation where guests are forced to take a looooong way around to get to their destination:

pathfinding5

The algorithm to set wide path flags may require a look-over.

### Steps to reproduce

Build paths in one of the above configurations. May depend on the orientation of the map.

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Effectively reverts 86be86331e53eb7c3a29b6d1d58cc47ab2ada27a. Should fix #22824.

Question remains what this code was originally aiming to _do_. The commit that originally changed the behaviour seems to suggest the function wouldn't run if it had been passed _tile_ coordinates, rather than the world coordinates it expects. That doesn't seem to be the case, though, as it checks against the lower 0xE0 bits having been set, which is the case every 224/32=7 tiles. It seems to be a means to prevent entire lengths of paths being cut off from the pathfinding algorithm.

In OpenRCT2, the check was evidently changed by 86be86331e53eb7c3a29b6d1d58cc47ab2ada27a and eventually removed entirely. This PR restores the behaviour, but the resulting 'wide path' flags do appear more erratic now. Clearly, this makes the pathfinding more flexible, but it warrants some thorough playtesting before this is considered for merging. Oddly enough, our replay tests appear to be fine, though. (Perhaps as a result of players working around the pathfinding oddities in the first place?)

@deurklink @LordMarcel Could you test this PR and report your findings? Builds can be downloaded from the 'Checks' tab, under 'CI'.

Before:
Test Park 2024-09-26 10-04-54

After:
Test Park 2024-09-26 10-05-26

### CODE REVIEW

File: src/openrct2/world/Footpath.cpp [hunk 1]
```diff
@@ -1596,6 +1596,11 @@ void FootpathUpdatePathWideFlags(const CoordsXY& footpathPos)
     // FootpathClearWide(x, y);
     // y -= 0x20;
 
+    // Check if tile coords are being passed? Why does that even happen?
+    // TODO: Needs adjusting for higher coords (NSF) before this is merged.
```
Comment (AaronVanGeffen) [inline src/openrct2/world/Footpath.cpp:1600]: Outdated comment; see OP.

**Comment by duncanspumpkin** [comment]
> Effectively reverts [86be863](https://github.com/OpenRCT2/OpenRCT2/commit/86be86331e53eb7c3a29b6d1d58cc47ab2ada27a). Should fix #22824.
> 
> Question remains what this code was originally aiming to _do_. The commit that originally changed the behaviour seems to suggest the function wouldn't run if it had been passed _tile_ coordinates, rather than the world coordinates it expects. That doesn't seem to be the case, though, as it checks against the lower 0xE0 bits having been set, which is the case every 224/32=7 tiles. It seems to be a means to prevent entire lengths of paths being cut off from the pathfinding algorithm.
> 
> In OpenRCT2, the check was evidently changed by [86be863](https://github.com/OpenRCT2/OpenRCT2/commit/86be86331e53eb7c3a29b6d1d58cc47ab2ada27a) and eventually removed entirely. This PR restores the behaviour, but the resulting 'wide path' flags do appear more erratic now. Clearly, this makes the pathfinding more flexible, but it warrants some thorough playtesting before this is considered for merging. Oddly enough, our replay tests appear to be fine, though. (Perhaps as a result of players working around the pathfinding oddities in the first place?)
> 
> @deurklink @LordMarcel Could you test this PR and report your findings? Builds can be downloaded from the 'Checks' tab, under 'CI'.

Bpb probably doesn't have wide plazas so that is probably why there is no change on the replays

**Comment by Gymnasiast** [comment]
The flag not being set every seven tiles rings a bell, actually. I seem to remember seeing that when we had just introduced wide path debugging.

**Comment by AaronVanGeffen** [comment]
I was wondering about the delay in recalculating the wide path flags. It's iterative on purpose, of course, but there's room for a good speedup:
```cpp
void MapUpdatePathWideFlags()
{
    // [...]
    CoordsXY& loopPosition = GetGameState().WidePathTileLoopPosition;
    for (int32_t i = 0; i < 128; i++)
    {
        FootpathUpdatePathWideFlags(loopPosition);

        // Next x, y tile
        loopPosition.x += kCoordsXYStep;
        if (loopPosition.x >= MAXIMUM_MAP_SIZE_BIG)
        {
            loopPosition.x = 0;
            loopPosition.y += kCoordsXYStep;
            if (loopPosition.y >= MAXIMUM_MAP_SIZE_BIG)
            {
                loopPosition.y = 0;
            }
        }
    }
}
```

Consider how `MAXIMUM_MAP_SIZE_BIG` has become much bigger with the introduction of the NSF. In fact, it only needs to consider the actual map size — and only the actual X and Y size at that. I'll file a separate PR to tackle that.

**Comment by sjoerddebruin** [comment]
Tested on a few parks, you can clearly see the effects on for example the parking of Over Texas.

image

After a few mins it turned into this on the old version

image

With these changes said issues don't appear.

**Comment by sjoerddebruin** [comment]
Crazy Castle still seems to have issues in this part:

image
_Before_

image
_After_

Guests are stuck in both cases.

File: distribution/changelog.txt [hunk 2]
```diff
@@ -8,6 +8,7 @@
 - Change: [#22740] Add virtual floor to shifted track design placement.
 - Change: [#22795] Replace Giga Coaster and Twister Roller Coaster booster images.
 - Fix: [#2614] The colour tab of the ride window does not hide invisible cars (original bug).
+- Fix: [#7672] Wide path status is set to all 'wide' paths, instead of only a quarter, impeding pathfinding.
```
Comment (AaronVanGeffen) [inline distribution/changelog.txt:11]: Quickly, replace these with typewriter quotes before @Gymnasiast finds out.
```suggestion
- Fix: [#7672] Wide path status is set to all ‘wide’ paths, instead of only a quarter, impeding pathfinding.
```

**Comment by deurklink** [comment]
I tested this PR in Imagination Megapark and Deurklink Megapark, both parks are around 254 x 254 tiles big. I waited for the new PR to mark the wide paths in the new fashion. After the wide path were flagged, I sped up the game and checked for the amount of lost guests. In both parks I did not see an increase in number of guests who reported getting lost or not being able to find a ride in this PR.

In Imagination Megapark I had some park squares that were impassable before this PR (so I had to modify them a bit to make them passable). With this PR, i could make the path squares full paths again and guests would traverse them just fine. 
tl;dr Seems to work fine in my biggest parks!

Impassable square before:
image

Square in this PR:
image

**Comment by deurklink** [comment]
The multiplayer park from #7672 now also handles the middle path much more nicely. It used to have a long impassable path running from one side to the other in the middle, but now it's more grid-like and allows guests to pass more easily.

image

File: src/openrct2/world/Footpath.cpp [hunk 3]
```diff
@@ -1596,6 +1596,11 @@ void FootpathUpdatePathWideFlags(const CoordsXY& footpathPos)
     // FootpathClearWide(x, y);
     // y -= 0x20;
 
+    // Only consider approx. 1/4 of tiles for wide path status
```
Comment (Gymnasiast) [inline src/openrct2/world/Footpath.cpp:1599]: Are you sure about these fractions? I would have expected something like 1/7 here.

Comment (AaronVanGeffen) [reply src/openrct2/world/Footpath.cpp:1599]: 0xE0 is 224, so technically it's 224/1000 tiles. I was rounding off; I think it's meant to approximate 25%, yeah.

Comment (AaronVanGeffen) [reply src/openrct2/world/Footpath.cpp:1599]: Argh no, it's hexadecimal, so (0xFF - 0xE0) / 0xFF ~= 0.12156, so much closer to 12.5%. You're right.

Comment (Gymnasiast) [reply src/openrct2/world/Footpath.cpp:1599]: I think that may be off by one - there 256 (0x100) different coordinates in a byte, from 0x0 to 0xFF, so I think the calculation is (0x100 - 0xE0) / 0x100, which neatly translates to 256.

Just for comparison, if you’d compare it with 0xFF instead of 0xE0, you’d still hit 0xFF once. In my calculation that’s (0x100 - 0xFF) / 0x100, = 1/256 whereas that would be 0 in your calculation.

**Review by Gymnasiast — APPROVED** [review]
Ship it!

**Comment by zaxcav** [comment]
> The flag not being set every seven tiles rings a bell, actually. I seem to remember seeing that when we had just introduced wide path debugging.

Yes, if I remember correctly, the idea was to define a default NxN tile grid of paths, that are not thinned by the wide path flags, to fill in wide paths (e.g. with crossing points), "courtyards" etc.
Amazing that the path finding still uses this after all these years. Keep up the good work!
